channeldb/invoices.go

```diff
@@ -2231,6 +2214,53 @@ func (d *DB) addHTLCs(invoices, settleIndex, //nolint:funlen
 		invoice.AmtPaid += amtPaid
 	}
 
+	err := d.addHTLCsStoreUpdate(
+		invoices, settleIndex, setIDIndex, invoiceNum, invoice,
+		settledSetIDs, htlcsAmpUpdate, timestamp,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return invoice, nil
+}
+
+func (d *DB) addHTLCsStoreUpdate(invoices, settleIndex, setIDIndex kvdb.RwBucket,
```

Comment:
Cool, I think this is inline with what I had in my pending mental model. Can be abstracted one layer further as well, since rn it gives it raw access to the various buckets. I'm thinking we decouple the description of the update, from the application itself. Under that mental model, this function is the application, with the set of arguments making up part of the description. The gap here then is taking a method call and mapping it to some description, to be applied atomically at a later step. 

One way to sort feel out these set of changes father would be to remove the `Invoice` argument from most/all of the update methods. Then the set of methods would be operating on an abstract `InvoiceUpdate` interface, which collects the updates, then applies them. For KV land application can just be mutate the invoice then sync to disk.

Reply:
I was also thinking this would be one layer higher. So most of this biz logic is now in the `invoice` package, then the idealized `channeldb` package is just doing the update (tho it is the case that rn we have a lot of biz logic in `chanenldb`, and part of the original `UpdateInvoice` call back was meant to help us start to extract that logic.

Reply:
Please note that the first few commits are just preliminary refactors in order to introduce the `InvoiceUpdater` and also to decouple in-memory invoice updates and DB updates. I should have flagged this in the PR description, but my general idea to do the least intrusive changes was to:

1) separate DB updates to their own functions
2) introduce a common interface and KV implementation for the updates
3) move all the update invoice logic to the `invoices` package
4) move all non-kv related tests to the `invoices` package too

This way we can implement an updater that's compatible with the existing implementation and works on SQL. The challenge here imho is the right level of abstraction as well as being careful with any changes to the `Invoice` or the in-memory update logic itself since our test coverage is somewhat limited. My choice here was to first make things work with a SQL updater in a naive way, and once we can deprecate the KV invoice DB do a deeper refactor to have a more SQL friendly implementation.

I'm looking into your idea of a more streamlined approach as I agree that having helpful types and more interfaces can generally help.

Reply:
In the mean time I'm working on the separate branch in parallel to actually implement the missing bits for the SQL port, so this PR sometimes changes as I discover simpler or more elegant ways to separate things.